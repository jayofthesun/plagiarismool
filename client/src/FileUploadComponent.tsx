import { Input, WithStyles, withStyles } from "@material-ui/core";
import React from "react";

/**
 * Props for the file upload component
 * @member filesCallback function to pass uploaded files back to the parent component
 */
interface IProps extends WithStyles<typeof styles> {
  filesCallback: (files: File[]) => any;
}

/**
 * States for the file upload component
 * @member selectedFiles
 */
interface IState {
  selectedFile: File[]
}


/**
 * styles for this component
 */
const styles = {
  container: {
    height: '350px',
    background: 'white',
    padding: '15px',
    borderRadius: '5px',
    margin: '15px',
    overflow: 'scroll'
  }
};

/**
 * Component that represents the file groups upload button
 */
export class FileUploadComponent extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props)
    this.state = { selectedFile: [] };
  }

  /**
   * Updates the current uploaded files
   * @param event the targeted input element to get the files
   */
  handleChange(event: any): void {
    var newFiles: File[] = Array.from(event.target.files);
    this.setState({ selectedFile: newFiles });
    this.props.filesCallback(newFiles)
  }

  /**
   * Renders the data of the current uploaded files
   */
  fileData() {
    if (this.state.selectedFile.length !== 0) {
      return (
        <div>
          <h2>Uploaded files:</h2>
          {Array.from(this.state.selectedFile).map((file: File, index: number) => {
            if (index === this.state.selectedFile.length - 1) {
              return (
                <span className="file-name" key={file.name + index}>{file.name}</span>
              )
            } else {
              return (
                <span className="file-name" key={file.name + index}>{file.name}, </span>
              )
            }
          })}
        </div>
      );
    }
  };

  render() {
    return <div className={this.props.classes.container}>
      <Input type="file" inputProps={{ multiple: true, accept: ".js, .jsx, .ts, .tsx" }} onChange={this.handleChange.bind(this)} />
      {this.fileData()}
    </div>
  }
}


export default withStyles(styles)(FileUploadComponent);