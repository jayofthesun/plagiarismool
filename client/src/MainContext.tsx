import React from 'react';
/**
 * Creates a global context to pass uploaded files between the upload component and the result component
 */
const defaultVal = {
  fileGroups: [] as File[][]
}

export const MainContext = React.createContext(defaultVal);
