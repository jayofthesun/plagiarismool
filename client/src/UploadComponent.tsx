
import React from "react";
import FileUploadComponent from './FileUploadComponent';
import { Button, Grid, WithStyles, withStyles } from '@material-ui/core';
import { MainContext } from "./MainContext";
import { RouteComponentProps, withRouter } from "react-router-dom";
import { DataService } from "./DataService";
import LoadingOverlay from 'react-loading-overlay';

/**
 * Props for the upload component
 */
interface IProps extends RouteComponentProps, WithStyles<typeof styles> { }

/**
 * State for the upload component
 * @member uploadedFiles pair of file groups the user uploaded
 * @member disabledButton checks if the button should be enabled or disabled
 * @member isLoading checks whether or not the loading bar should shows up
 */
interface IState {
  uploadedFiles: File[][];
  disabledButton: boolean;
  isLoading: boolean;
}

/**
 * styles for this component
 */
const styles = {
  title: {
    marginTop: '2%',
    marginLeft: '4%'
  },
  gridContainer: {
    padding: '2% 4%'
  }
};

/**
 * The upload component page, the controller for the upload side of this react app
 */
export class UploadComponent extends React.Component<IProps, IState> {
  static contextType = MainContext;

  constructor(props: IProps) {
    super(props);
    this.state = {
      uploadedFiles: [[], []],
      disabledButton: true,
      isLoading: false
      
    };
  }

  /**
   * Sets the state of the selected files specfically on the upload page.
   * 
   * @param file The files that are being uploaded.
   */
  updateFiles(values: File[], index: number) {
    const newUploadedFiles = this.state.uploadedFiles;
    newUploadedFiles[index] = values;
    this.setState({ uploadedFiles: newUploadedFiles });
    this.updateButton();
  }

  /**
   * Pushes the updated state to the list
   * Since setting state is asyncronous, this has to be a separate function that launches post state set.
   */
  updateButton() {
    var hasItems = true;
    for (const fileGroup of this.state.uploadedFiles) {
      hasItems = ((fileGroup.length !== 0) && hasItems);
    }
    this.setState({ disabledButton: !hasItems })
  }

  /**
   * Sets the data of files into a formdata to send to API endpoint
   */
  onFileUpload() {
    this.setState({ isLoading: true })

    // Create an object of formData 
    if (this.state.uploadedFiles[0] === undefined || this.state.uploadedFiles[1] === undefined) {
      throw new Error("Invalid file upload")
    } else {
      const formData = new FormData();
      const newFileList1: File[] = [];
      const newFileList2: File[] = [];

      //For list 1
      for (let i = 0; i < this.state.uploadedFiles[0].length; i++) {
        const currentFile = this.state.uploadedFiles[0][i];
        formData.append(
          "program1",
          currentFile,
          currentFile.name
        );
        newFileList1.push(currentFile)

      }
      //For list 2
      for (let i = 0; i < this.state.uploadedFiles[1].length; i++) {
        const currentFile = this.state.uploadedFiles[1][i];
        formData.append(
          "program2",
          currentFile,
          currentFile.name
        );
        newFileList2.push(currentFile)
      }

      //API end point call
      DataService.addFiles(formData).then((response: any) => {
        const newFileLists = [newFileList1, newFileList2]
        this.context.setFileGroups(newFileLists);
        this.setState({ isLoading: false })
        this.props.history.push('/results');
      });
    }
  };


  render() {
    return <LoadingOverlay
      active={this.state.isLoading}
      spinner
      styles={{ wrapper: { position: "inherit" } }}
      text='Hang on...your files are being scanned!'>
      <section className="upload-body">
        <div className={this.props.classes.title}>
          <h1>Typescript Plagiarism Detector</h1>
          <p>Helps detect potential plagiarism files for .js, .jsx, .ts, and .tsx files</p>
        </div>
        <Grid
          container
          direction="row"
          justify="space-evenly"
          className={this.props.classes.gridContainer}
        >
          <Grid item xs={12}>
            <h3>First, upload your two groups of files:</h3>
          </Grid>
          <Grid item xs={5}>
            <FileUploadComponent filesCallback={(files: File[]) => this.updateFiles(files, 0)} />
          </Grid>
          <Grid item xs={5}>
            <FileUploadComponent filesCallback={(files: File[]) => this.updateFiles(files, 1)} />
          </Grid>
          <Grid item xs={12}>
            <h3>Next, start scan!</h3>
          </Grid>
          <Grid item xs={12}>
            <Button variant="contained"
              color="primary"
              disabled={this.state.disabledButton}
              onClick={this.onFileUpload.bind(this)}>
              Start Scan!
          </Button>
          </Grid>
        </Grid>
      </section >
    </LoadingOverlay>
  }
}

export default withStyles(styles)(withRouter(UploadComponent));