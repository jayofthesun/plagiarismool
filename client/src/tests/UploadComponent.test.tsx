import React from 'react';
import UploadComponent, { UploadComponent as OriginalComponent } from '../UploadComponent';
import { Button, Input } from '@material-ui/core';
import { ReactWrapper, shallow } from 'enzyme';
import { FileUploadComponent as OriginalUploadComponent } from '../FileUploadComponent';
import { Router } from 'react-router';
import { MainContext } from '../MainContext';
import { createMemoryHistory } from 'history'
import { createMount } from '@material-ui/core/test-utils';
import { render } from '@testing-library/react';

Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
    })),
});

describe('tests for the main upload component', () => {
    //Dummy files
    const file = new File(["hello", "oh boy"], "cool.ts");
    const file2 = new File(["coolio"], "coolio.ts");
    const file3 = new File(["sample"], 'sample.ts', {});
    const file4 = new File(["hello", "oh boy"], 'sample2.ts', {});
    const fileList = [[file, file2], [file3, file4]];

    //main context mock filegroups
    const contextFileGroups = {
        fileGroups: fileList,
        setFileGroups: ((files: File[][]) => undefined)
    }

    //routing mock history
    const history = createMemoryHistory({
        initialEntries: ['/'],
    });
    let mount;

    let wrapper: ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;
    beforeEach(() => {

        mount = createMount();
        wrapper = mount(<MainContext.Provider value={contextFileGroups}>
            <Router history={history}>
                <UploadComponent />
            </Router>
        </MainContext.Provider>);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('UploadComponent snapshot is properly rendered.', () => {
        const component = render(<Router history={history}>
            <UploadComponent />
        </Router>)
        expect(component).toMatchSnapshot();
    })

    it('make sure visual elements render, starting case', () => {
        const component = wrapper.find(OriginalComponent)
        expect(component.find(OriginalUploadComponent), "sees if inputs component exists").toHaveLength(2);
        expect(component.find(Input), "sees if inputs exists").toHaveLength(2);
        expect(component.find(Button).props().disabled, "sees if button is disabled").toBe(true);
    });

    it('uploading one file to the left side', () => {
        let component = wrapper.find(OriginalComponent)
        expect(component.find(Button).props().disabled, "sees if button is disabled").toBe(true);
        component.find('input').at(0).simulate('change', {
            target: {
                files: [
                    file
                ]
            }
        });
        expect(component.find(Button).props().disabled, "sees if button is still disabled").toBe(true);
    });

    it('uploading one file to the right side', () => {
        let component = wrapper.find(OriginalComponent)
        expect(component.find(Button).props().disabled, "sees if button is disabled").toBe(true);
        component.find('input').at(1).simulate('change', {
            target: {
                files: [
                    file
                ]
            }
        });
        expect(component.find(Button).props().disabled, "sees if button is still disabled").toBe(true);
    });

    it('uploading one file to the both sides', () => {
        let component = wrapper.find(OriginalComponent)
        expect(component.find(Button).props().disabled, "sees if button is disabled").toBe(true);
        component.find('input').at(0).simulate('change', {
            target: {
                files: [
                    file
                ]
            }
        });

        component = wrapper.find(OriginalComponent)
        expect(component.find(Button).props().disabled, "sees if button is still disabled").toBe(true);
        component.find('input').at(1).simulate('change', {
            target: {
                files: [
                    file
                ]
            }
        });

        component = wrapper.find(OriginalComponent)
        expect(component.find(Button).props().disabled, "sees if button is not disabled").toBe(false);
    });

    it('click on uploaded button to get to result page', () => {
        const component = wrapper.find(OriginalComponent)
        expect(component.find(Input), "sees if input exists").toHaveLength(2);
        expect(component.find(Button).prop('disabled'), "sees if button is disabled").toBe(true);
        expect(component.find(".file-name"), "there should be no file names when starting").toHaveLength(0);
    });

    it('if you click on upload button loading screen works', () => {
        let component = wrapper.find(OriginalComponent)
        expect(component.state().isLoading, "see if screen is not loading - disabled button").toBe(false);
        expect(component.find('LoadingOverlayWrapper').prop('active'), "see if screen is not loading default - disabled button").toBe(false);
        component.find(Button).simulate('click', { button: 0 }); // try to click on disabled button

        component = wrapper.find(OriginalComponent)
        expect(component.find('LoadingOverlayWrapper').prop('active'), "see if screen is still loading - clicking disabled button works").toBe(false);

        //enable button clicking
        component.setState({ disabledButton: false })
        component = wrapper.find(OriginalComponent)
        expect(component.find('LoadingOverlayWrapper').prop('active'), "see if screen is not loading still").toBe(false);
        component.find(Button).simulate('click', { button: 0 });
        component = wrapper.find(OriginalComponent)
        expect(component.find('LoadingOverlayWrapper').prop('active'), "see if screen is loading when clicked").toBe(true);
    });
});