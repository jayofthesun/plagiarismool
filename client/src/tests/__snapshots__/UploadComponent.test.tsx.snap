// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for the main upload component UploadComponent snapshot is properly rendered. 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      class="app"
      id="app"
    >
      <div
        class="_loading_overlay_wrapper css-18nsqm8"
        data-testid="wrapper"
      >
        <section
          class="upload-body"
        >
          <div
            class="withRouter(UploadComponent)-title-1"
          >
            <h1>
              Typescript Plagiarism Detector
            </h1>
            <p>
              Helps detect potential plagiarism files for .js, .jsx, .ts, and .tsx files
            </p>
          </div>
          <div
            class="MuiGrid-root withRouter(UploadComponent)-gridContainer-2 MuiGrid-container MuiGrid-justify-xs-space-evenly"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <h3>
                First, upload your two groups of files:
              </h3>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
            >
              <div
                class="FileUploadComponent-container-3"
              >
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline"
                >
                  <input
                    accept=".js, .jsx, .ts, .tsx"
                    class="MuiInputBase-input MuiInput-input"
                    multiple=""
                    type="file"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
            >
              <div
                class="FileUploadComponent-container-3"
              >
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline"
                >
                  <input
                    accept=".js, .jsx, .ts, .tsx"
                    class="MuiInputBase-input MuiInput-input"
                    multiple=""
                    type="file"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <h3>
                Next, start scan!
              </h3>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Start Scan!
                </span>
              </button>
            </div>
          </div>
        </section>
      </div>
    </div>
    <div>
      <div
        class="_loading_overlay_wrapper css-18nsqm8"
        data-testid="wrapper"
      >
        <section
          class="upload-body"
        >
          <div
            class="withRouter(UploadComponent)-title-1"
          >
            <h1>
              Typescript Plagiarism Detector
            </h1>
            <p>
              Helps detect potential plagiarism files for .js, .jsx, .ts, and .tsx files
            </p>
          </div>
          <div
            class="MuiGrid-root withRouter(UploadComponent)-gridContainer-2 MuiGrid-container MuiGrid-justify-xs-space-evenly"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <h3>
                First, upload your two groups of files:
              </h3>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
            >
              <div
                class="FileUploadComponent-container-3"
              >
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline"
                >
                  <input
                    accept=".js, .jsx, .ts, .tsx"
                    class="MuiInputBase-input MuiInput-input"
                    multiple=""
                    type="file"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
            >
              <div
                class="FileUploadComponent-container-3"
              >
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline"
                >
                  <input
                    accept=".js, .jsx, .ts, .tsx"
                    class="MuiInputBase-input MuiInput-input"
                    multiple=""
                    type="file"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <h3>
                Next, start scan!
              </h3>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Start Scan!
                </span>
              </button>
            </div>
          </div>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_loading_overlay_wrapper css-18nsqm8"
      data-testid="wrapper"
    >
      <section
        class="upload-body"
      >
        <div
          class="withRouter(UploadComponent)-title-1"
        >
          <h1>
            Typescript Plagiarism Detector
          </h1>
          <p>
            Helps detect potential plagiarism files for .js, .jsx, .ts, and .tsx files
          </p>
        </div>
        <div
          class="MuiGrid-root withRouter(UploadComponent)-gridContainer-2 MuiGrid-container MuiGrid-justify-xs-space-evenly"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <h3>
              First, upload your two groups of files:
            </h3>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
          >
            <div
              class="FileUploadComponent-container-3"
            >
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline"
              >
                <input
                  accept=".js, .jsx, .ts, .tsx"
                  class="MuiInputBase-input MuiInput-input"
                  multiple=""
                  type="file"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
          >
            <div
              class="FileUploadComponent-container-3"
            >
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline"
              >
                <input
                  accept=".js, .jsx, .ts, .tsx"
                  class="MuiInputBase-input MuiInput-input"
                  multiple=""
                  type="file"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <h3>
              Next, start scan!
            </h3>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Start Scan!
              </span>
            </button>
          </div>
        </div>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
